### PR TITLE
Implementing a buffer that handles dataset elements of different sizes

### PR DESCRIPTION
**Pull request by prabhuteja12**
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/renate/memory/test_storage.py [hunk 1]
```diff
@@ -21,25 +22,72 @@ def nested_tensors_equal(t1, t2):
...
 
     del storage
-    storage = MemoryMappedTensorStorage(tmpdir, data_point, length)
+    storage = StorageClass(tmpdir)
+    for i in range(length):
+        assert nested_tensors_equal(storage[i], ds[i])
 
+
+def make_dataset_different_sizes(dataset_length, return_type):
+    class CustomDataset2(Dataset):
```
Comment (wistuba) [inline test/renate/memory/test_storage.py:63]: nit: any reason to use "2" here?

File: test/renate/memory/test_storage.py [hunk 2]
```diff
@@ -21,25 +22,72 @@ def nested_tensors_equal(t1, t2):
...
+    class CustomDataset(Dataset):
+        def __init__(self, x, y, return_type="tensor") -> None:
+            self.x, self.y = x, y
+            self.return_type = return_type
+
+        def __getitem__(self, index):
+            if self.return_type == "tensor":
+                return self.x[index]
+            elif self.return_type == "tuple":
+                return (self.x[index], self.y[index])
```
Comment (wistuba) [inline test/renate/memory/test_storage.py:35]: why are here brackets?

File: test/renate/memory/test_storage.py [hunk 3]
```diff
@@ -21,25 +22,72 @@ def nested_tensors_equal(t1, t2):
...
+def make_dataset_different_sizes(dataset_length, return_type):
+    class CustomDataset2(Dataset):
+        def __init__(self, return_type="tensor") -> None:
+            self.return_type = return_type
+
+        def __getitem__(self, index):
+            x = torch.ones(index + 2, index + 3).float() * (index + 1)
+            y = torch.ones(index + 2, index + 3).int() * (index + 2)
+            if self.return_type == "tensor":
+                return x, y
```
Comment (wistuba) [inline test/renate/memory/test_storage.py:71]: how is this different from "tuple"?

File: test/renate/memory/test_storage.py [hunk 4]
```diff
@@ -21,25 +22,72 @@ def nested_tensors_equal(t1, t2):
...
+
+    data_x = torch.rand(dataset_length, 3, 32, 32)
+    data_y = torch.randint_like(data_x, 5)
+
+    return CustomDataset(data_x, data_y, return_type=return_type)
+
+
+@pytest.mark.parametrize("length", [1, 10])
+@pytest.mark.parametrize("return_type", ["tuple", "dict", "tensor"])
+@pytest.mark.parametrize("StorageClass", [FileTensorStorage])
```
Comment (wistuba) [inline test/renate/memory/test_storage.py:50]: is this needed? you don't have it in the other test?

File: test/renate/memory/test_storage.py [hunk 5]
```diff
@@ -21,25 +22,72 @@ def nested_tensors_equal(t1, t2):
...
+def test_memory_storage_same_size(tmpdir, length, return_type, StorageClass):
+    ds = make_dataset_same_sizes(length, return_type)
+    storage = StorageClass(tmpdir)
+    storage.dump_dataset(ds)
 
     del storage
-    storage = MemoryMappedTensorStorage(tmpdir, data_point, length)
+    storage = StorageClass(tmpdir)
+    for i in range(length):
+        assert nested_tensors_equal(storage[i], ds[i])
```
Comment (wistuba) [inline test/renate/memory/test_storage.py:59]: test logic is identical to test logic for "different size" besides the very first line. reuse validation logic

File: test/renate/memory/test_storage.py [hunk 6]
```diff
@@ -2,16 +2,17 @@
...
+from renate.memory.storage import FileTensorStorage
 
 
 def nested_tensors_equal(t1, t2):
     if type(t1) is not type(t2):
         return False
     if isinstance(t1, torch.Tensor):
         return torch.equal(t1, t2)
-    if isinstance(t1, tuple):
+    if isinstance(t1, (tuple, list)):
```
Comment (wistuba) [inline test/renate/memory/test_storage.py:15]: are we testing for lists?

File: src/renate/memory/storage.py [hunk 7]
```diff
@@ -111,6 +129,40 @@ def _set(storage: NestedTensors, idx: int, data_point: NestedTensors) -> None:
         else:
             raise TypeError(f"Expected nested tuple/dict of tensors, found {type(storage)}.")
 
-    def __setitem__(self, idx: int, data_point: NestedTensors) -> None:
-        """Set the item stored at index `idx`."""
-        self._set(self._storage, idx, data_point)
+    def dump_dataset(self, ds):
+        self._length = len(ds)
```
Comment (wistuba) [inline src/renate/memory/storage.py:133]: is `self._length` needed?

File: src/renate/memory/storage.py [hunk 8]
```diff
@@ -50,9 +61,16 @@ class MemoryMappedTensorStorage(Storage):
         length: Number of items to be stored.
     """
 
-    def __init__(self, directory: str, data_point: NestedTensors, length: int) -> None:
-        super().__init__(directory, data_point, length)
-        self._storage = self._create_mmap_tensors(directory, data_point, length)
+    def __init__(self, directory: str) -> None:
```
Comment (wistuba) [inline src/renate/memory/storage.py:64]: why do we keep this class?

File: src/renate/memory/storage.py [hunk 9]
```diff
@@ -111,6 +129,40 @@ def _set(storage: NestedTensors, idx: int, data_point: NestedTensors) -> None:
...
+        directory: Path to a directory.
+    """
+
+    def __init__(self, directory: str) -> None:
+        super().__init__(directory)
+
+    def dump_dataset(self, ds: torch.utils.data.Dataset) -> None:
+        for i in range(len(ds)):
+            ith_sample = ds[i]
+            save_name = os.path.join(self._directory, f"{i}.pt")
```
Comment (wistuba) [inline src/renate/memory/storage.py:160]: maybe an overkill, but one could make this a function

**Review by wistuba — CHANGES_REQUESTED** [review]

File: test/renate/memory/test_storage.py (hunk 3 above)
Comment (prabhuteja12) [reply test/renate/memory/test_storage.py:71]: Fixed.

File: test/renate/memory/test_storage.py (hunk 4 above)
Comment (prabhuteja12) [reply test/renate/memory/test_storage.py:50]: merged both tests.

File: test/renate/memory/test_storage.py (hunk 2 above)
Comment (prabhuteja12) [reply test/renate/memory/test_storage.py:35]: differentiating from returning list.

File: src/renate/memory/storage.py (hunk 8 above)
Comment (prabhuteja12) [reply src/renate/memory/storage.py:64]: Purely for reference. I suggest we keep this class even though its not used anywhere.

File: src/renate/memory/storage.py (hunk 7 above)
Comment (prabhuteja12) [reply src/renate/memory/storage.py:133]: Yes. This is returned in __len__ call

File: src/renate/memory/storage.py [hunk 10]
```diff
@@ -111,6 +129,40 @@ def _set(storage: NestedTensors, idx: int, data_point: NestedTensors) -> None:
...
+    Args:
+        directory: Path to a directory.
+    """
+
+    def __init__(self, directory: str) -> None:
+        super().__init__(directory)
+
+    def dump_dataset(self, ds: torch.utils.data.Dataset) -> None:
+        for i in range(len(ds)):
+            ith_sample = ds[i]
```
[the remark this answers is not in the document]
Comment (prabhuteja12) [reply src/renate/memory/storage.py:159]: done.

File: src/renate/memory/storage.py (hunk 9 above)
Comment (prabhuteja12) [reply src/renate/memory/storage.py:160]: done.

**Pull request merged**